src/qutip_qip/decompose/decompose_general_qubit_gate.py

```diff
@@ -0,0 +1,932 @@
+    for i in range(len_two_level):
+        index1, index2 = two_level_array_output[i][0]
+        gray_code_info = _gray_code_gate_info(index1, index2, num_qubits)
+        if len(gray_code_info) == 1:
+            single_step.append(i)
+
+    # find the index with target_qubit = 0
+    for i in range(len(single_step)):
+        ind = single_step[i]
+        index1, index2 = two_level_array_output[ind][0]
+        gray_code = _gray_code_gate_info(index1, index2, num_qubits)
+        gray_code_target = gray_code[0]['targets =']
+        if gray_code_target[0] == 0:
+            return(ind)
+
+
+def _full_CNOT_to_Toffoli(num_qubits):
+    """Decomposes a multi-controlled CNOT with last qubit as target into a
+    circuit composed of toffoli and ancilla qubits.
+    """
+    if num_qubits >= 3:
+        num_ctrl_qubits = num_qubits - 1
+        num_ancilla_qubits = num_ctrl_qubits - 1
+        # change last qubit due to ancillas
+        last_target_qubit = num_ctrl_qubits + num_ancilla_qubits
+
+        # create ctrl qubit list
+        ctrl_qubit_list = list(range(num_ctrl_qubits))
+        # create ancilla qubit list
+        ancilla_qubit_list = list(range(num_ctrl_qubits, last_target_qubit))
+
+        # first toffoli uses first 2 qubits and first ancilla
+        Toffoli0 = Gate(
+            "TOFFOLI", targets=[ancilla_qubit_list[0]], controls=[0, 1])
+        # remove ancilla qubit and ctrl qubits from their respective lists
+        ctrl_qubit_list.remove(0)
+        ctrl_qubit_list.remove(1)
+
+        # create a list of control and target qubits for toffoli gates
+        ctrls_targ_list = []
+        for i in range(len(ancilla_qubit_list)-1):
+            # index for target qubits
+            j = i + 1
+            ctrls_at_i = [ctrl_qubit_list[i], ancilla_qubit_list[i]]
+            targ_at_i = [ancilla_qubit_list[j]]
+            ctrls_targ_list.append([ctrls_at_i, targ_at_i])
+
+        # create list of toffoli gates
+        full_toffoli_list = [Toffoli0]
+        for i in range(len(ctrls_targ_list)):
+            ctrls_i = ctrls_targ_list[i][0]
+            targ_i = ctrls_targ_list[i][1]
+            toffoli_at_i = Gate("TOFFOLI", targets=targ_i, controls=ctrls_i)
+            full_toffoli_list.append(toffoli_at_i)
+
+        # middle CNOT
+        CNOT_middle = Gate(
+            "CNOT", controls=ancilla_qubit_list[-1], targets=last_target_qubit)
+
+        # full gate list
+        full_gate_list = []
+        for i in range(len(full_toffoli_list)):
+            full_gate_list.append(full_toffoli_list[i])
+
+        full_gate_list.append(CNOT_middle)
+
+        for i in range(len(full_toffoli_list)):
+            reversed_list = full_toffoli_list[::-1]
+            full_gate_list.append(reversed_list[i])
+        return(full_gate_list)
+
+
+def _total_num_qubits_with_ancilla(num_qubits):
+    """Finds total number of qubits when ancilla qubits are added to create a
+    qubit circuit.
+    """
+    if num_qubits > 3:
```

Comment:
This needs to be changed to number of ctrl qubits - if n is the total number of qubits then total number of possible ctrl qubits are n-1.